examples/plugin/src/main.rs

```diff
@@ -0,0 +1,38 @@
+use plugin_api::plugin_api as pylib_module;
+use pyo3::prelude::*;
+use pyo3::types::PyList;
+use std::path::Path;
+
+fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    //"export" our API module to the python runtime
+    pyo3::append_to_inittab!(pylib_module);
+    //spawn runtime
+    pyo3::prepare_freethreaded_python();
+    //import path for python
+    let path = Path::new("./");
+    //do useful work
+    Python::with_gil(|py| {
+        //add the current directory to import path of Python (do not use this in production!)
+        let syspath: &PyList = py.import("sys")?.getattr("path")?.extract()?;
+        syspath.insert(0, &path)?;
+        println!("Import path is: {:?}", syspath);
+
+        // Now we can load our plugin.py file
+        let plugin = PyModule::import(py, "plugin")?;
+        // and call start function there, which will return a python reference to Gadget
+        let gadget = plugin.getattr("start")?.call0()?;
+        {
+            //now we extract the rust struct from python object
+            let mut gadget_rs: PyRefMut<'_,plugin_api::Gadget> = gadget.extract()?;
+            // we can now modify it as if it was a native rust struct
+            gadget_rs.prop = 42;
+            //which includes access to rust-only fields that are not visible to python
```

Comment:
This is an important advantage of this approach 👍 . You should put more emphasis on it. This entire example is about testing your python module *while* accessing its internal state, but that's not really mentioned anywhere else.

Reply:
Well not just testing, the reason I built this entire example is to figure out how to make Python scripting API in my application. Since my application is fairly complicated, I wanted to test certain pieces of the API separately from the main app. So access to internal state was a must. Also it gives ability to "hide" stuff from Python plugins that they do not need to access, while having access from Rust.